diff --git a/vsort/win32.cpp b/vsort/win32.cpp
index 0c7026e..e827c9a 100644
--- a/vsort/win32.cpp
+++ b/vsort/win32.cpp
@@ -27,6 +27,7 @@ static std::vector<std::wstring> cudaDlls {
     L"cudnn_graph64",
     L"cudnn_engines_precompiled64", L"cudnn_heuristic64", L"cudnn_engines_runtime_compiled64",
     L"cudnn_ops64", L"cudnn_cnn64", L"cudnn_adv64",
+    L"cudnn_engines_tensor_ir64", L"cudnn_ext64",
     L"cudnn64",
     L"cupti64",
 };
