diff --git a/vsncnn/onnx2ncnn.cpp b/vsncnn/onnx2ncnn.cpp
index ddf0d5e..be000bd 100644
--- a/vsncnn/onnx2ncnn.cpp
+++ b/vsncnn/onnx2ncnn.cpp
@@ -3592,6 +3592,18 @@ std::optional<std::tuple<char*, unsigned char*>> onnx2ncnn(onnx::ModelProto& mod
 
             std::string auto_pad = get_node_attr_s(node, "auto_pad");
             std::vector<int> kernel_shape = get_node_attr_ai(node, "kernel_shape");
+            if (kernel_shape.empty()) {
+                if (W.dims_size() == 3) {
+                    kernel_shape.push_back(W.dims(2));
+                } else if (W.dims_size() == 4) {
+                    kernel_shape.push_back(W.dims(2));
+                    kernel_shape.push_back(W.dims(3));
+                } else if (W.dims_size() == 5) {
+                    kernel_shape.push_back(W.dims(2));
+                    kernel_shape.push_back(W.dims(3));
+                    kernel_shape.push_back(W.dims(4));
+                }
+            }
             std::vector<int> dilations = get_node_attr_ai(node, "dilations");
             std::vector<int> strides = get_node_attr_ai(node, "strides");
             std::vector<int> pads = get_node_attr_ai(node, "pads");
@@ -3662,6 +3674,18 @@ std::optional<std::tuple<char*, unsigned char*>> onnx2ncnn(onnx::ModelProto& mod
 
             std::string auto_pad = get_node_attr_s(node, "auto_pad");
             std::vector<int> kernel_shape = get_node_attr_ai(node, "kernel_shape");
+            if (kernel_shape.empty()) {
+                if (W.dims_size() == 3) {
+                    kernel_shape.push_back(W.dims(2));
+                } else if (W.dims_size() == 4) {
+                    kernel_shape.push_back(W.dims(2));
+                    kernel_shape.push_back(W.dims(3));
+                } else if (W.dims_size() == 5) {
+                    kernel_shape.push_back(W.dims(2));
+                    kernel_shape.push_back(W.dims(3));
+                    kernel_shape.push_back(W.dims(4));
+                }
+            }
             std::vector<int> dilations = get_node_attr_ai(node, "dilations");
             std::vector<int> strides = get_node_attr_ai(node, "strides");
             std::vector<int> output_padding = get_node_attr_ai(node, "output_padding");
