diff --git a/vsncnn/vs_ncnn.cpp b/vsncnn/vs_ncnn.cpp
index 10434ce..9503f56 100644
--- a/vsncnn/vs_ncnn.cpp
+++ b/vsncnn/vs_ncnn.cpp
@@ -294,11 +294,8 @@ static const VSFrameRef *VS_CC vsNcnnGetFrame(
                 int x_crop_end = (x == src_width - src_tile_w) ? 0 : d->overlap_w;
 
                 {
-                    auto input_buffer = reinterpret_cast<uint8_t *>(
-                        d->fp16 && !fp16_input ?
-                        resource.h_src_fp32.data :
-                        resource.h_src.data
-                    );
+                    const auto & input_mat = d->fp16 && !fp16_input ? resource.h_src_fp32 : resource.h_src;
+                    auto input_buffer = reinterpret_cast<uint8_t *>(input_mat.data);
 
                     // assumes the pitches of ncnn::Mat to be
                     // (cstep * elemsize, w * h * elemsize, h * elemsize)
@@ -313,7 +310,7 @@ static const VSFrameRef *VS_CC vsNcnnGetFrame(
                                 src_ptr, src_stride,
                                 src_tile_w_bytes, src_tile_h
                             );
-                            input_buffer += resource.h_src.cstep * in_vis[0]->format->bytesPerSample;
+                            input_buffer += input_mat.cstep * input_mat.elemsize;
                         }
                     }
                 }
@@ -347,11 +344,8 @@ static const VSFrameRef *VS_CC vsNcnnGetFrame(
                 }
 
                 {
-                    auto output_buffer = reinterpret_cast<uint8_t *>(
-                        d->fp16 && !d->fp16_output ?
-                        resource.h_dst_fp32.data :
-                        resource.h_dst.data
-                    );
+                    const auto & output_mat = d->fp16 && !d->fp16_output ? resource.h_dst_fp32 : resource.h_dst;
+                    auto output_buffer = reinterpret_cast<uint8_t *>(output_mat.data);
 
                     for (int plane = 0; plane < dst_planes; ++plane) {
                         auto dst_ptr = (dst_ptrs[plane] +
@@ -368,7 +362,7 @@ static const VSFrameRef *VS_CC vsNcnnGetFrame(
                                 dst_tile_h - (y_crop_start + y_crop_end)
                             );
 
-                            output_buffer += resource.h_dst.cstep * d->out_vi->format->bytesPerSample;
+                            output_buffer += output_mat.cstep * output_mat.elemsize;
                         }
                     }
                 }
@@ -422,9 +416,17 @@ static void VS_CC vsNcnnFree(
         vsapi->freeNode(node);
     }
 
+    std::vector<std::pair<ncnn::VkAllocator *, ncnn::VkAllocator *>> allocators;
+    allocators.reserve(d->resources.size());
     for (const auto & resource : d->resources) {
-        d->device->reclaim_blob_allocator(resource.blob_vkallocator);
-        d->device->reclaim_staging_allocator(resource.staging_vkallocator);
+        allocators.emplace_back(resource.blob_vkallocator, resource.staging_vkallocator);
+    }
+
+    d->resources.clear();
+
+    for (auto [blob, staging] : allocators) {
+        d->device->reclaim_blob_allocator(blob);
+        d->device->reclaim_staging_allocator(staging);
     }
 
     delete d;
